### PR TITLE
[Rosetta] support stacking, delegate stacking and revoke stacking event

### PR DESCRIPTION
**Pull request by rllola**
## Description

This PR is adding to rosetta support for stacking, delegate stacking and revoke stacking events.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] API reference/documentation update
- [ ] Other

## Does this introduce a breaking change?
No

## Testing information
Tested using rosetta-cli

## Checklist
- [ ] Code is commented where needed
- [ ] Unit test coverage for new or modified code paths
- [ ] `npm run test` passes
- [ ] Changelog is updated
- [ ] Tag 1 of @kyranjamie or @zone117x for review

### CODE REVIEW

File: src/api/rosetta-constants.ts [hunk 1]
```diff
@@ -51,9 +51,15 @@ export const RosettaOperationTypes = [
...
   'miner_reward',
-  'stx_lock',
+  'stx_lock', // stx event
+  'stx_unlock', // forged event
+  'stack-stx', // PoX contract function
+  'delegate-stx', // PoX contract function
+  'revoke-delegate-stx', // PoX contract function
 ];
 
+export const PoxContractId = 'SP000000000000000000002Q6VF78.pox';
```
Comment (zone117x) [inline src/api/rosetta-constants.ts:61]: This should have a testnet and mainnet constant, similar to the bns usage here:
https://github.com/blockstack/stacks-blockchain-api/blob/0565d4f7f3130ed418b4652be3ae3c2a548bffe2/src/bns-constants.ts#L17-L20

File: src/datastore/postgres-store.ts [hunk 2]
```diff
@@ -4053,6 +4052,51 @@ export class PgDataStore
...
+      locked_amount: string;
+      unlock_height: string;
+      block_height: string;
+      locked_address: string;
+      tx_id: Buffer;
+    }>(
+      `
+      SELECT locked_amount, unlock_height, block_height, tx_id, locked_address
+      FROM stx_lock_events
+      WHERE canonical = true AND unlock_height = $1
```
Comment (zone117x) [inline src/datastore/postgres-store.ts:4075]: I'm not sure this query will always work as intended. There can be multiple burn blocks in between Stacks blocks. For example, given `(stacks block 1, burn block 1)` and `(stacks-block 2, burn block 4)`, if unlock height is at burn block 3, then the functions that calls this using a burn block associated with a stacks block will never return the unlock data at burn block 3.

It probably needs changed to find entries where the burn block is less than or equal to `stackBlock.burn_block_height` and greater than `previousStacksBlock.burn_block_height`.

File: src/datastore/postgres-store.ts [hunk 3]
```diff
@@ -1,64 +1,65 @@
 import * as path from 'path';
 import { EventEmitter } from 'events';
 import PgMigrate, { RunnerOption } from 'node-pg-migrate';
-import { Pool, PoolClient, ClientConfig, Client, ClientBase, QueryResult, QueryConfig } from 'pg';
```
Comment (zone117x) [inline src/datastore/postgres-store.ts:4]: Can all these formatting changes be reverted? I'd be in favor of adding an eslint rule to enforce alphanumeric ordering of imports, but that should be done in a separate PR where we apply it to the whole codebase. In this PR, it just creates noise and merge conflicts.

File: src/datastore/postgres-store.ts (hunk 2 above)
Comment (raynaudoe) [reply src/datastore/postgres-store.ts]: Yes, you're right. Fixed it.

File: src/datastore/postgres-store.ts (hunk 3 above)
Comment (raynaudoe) [reply src/datastore/postgres-store.ts:4]: Sure. Reverted

File: src/datastore/common.ts [hunk 4]
```diff
@@ -254,7 +255,20 @@ export interface DbNftEvent extends DbContractAssetEvent {
   value: Buffer;
 }
 
-export type DbEvent = DbSmartContractEvent | DbStxEvent | DbStxLockEvent | DbFtEvent | DbNftEvent;
+export interface StxUnlockEvent extends DbEventBase {
```
[the remark this answers is not in the document]
Comment (raynaudoe) [reply src/datastore/common.ts]: Ok, I've detached the definition pf `StxUnlockEvent` so it doesn't get confused as a `DBEvent` type.

File: src/api/rosetta-constants.ts [hunk 5]
```diff
@@ -51,9 +51,18 @@ export const RosettaOperationTypes = [
   'mint',
   'burn',
   'miner_reward',
-  'stx_lock',
+  'stx_lock', // stx event
+  'stx_unlock', // forged event
+  'stack-stx', // PoX contract function
+  'delegate-stx', // PoX contract function
+  'revoke-delegate-stx', // PoX contract function
```
Comment (zone117x) [inline src/api/rosetta-constants.ts:58]: Nit: convention diverges from underscore to hyphens here

File: src/api/rosetta-constants.ts [hunk 6]
```diff
@@ -51,9 +51,18 @@ export const RosettaOperationTypes = [
...
 ];
 
+export const enum BnsContractIdentifier {
```
Comment (asimm241) [inline src/api/rosetta-constants.ts:61]: These constants are already declared under `bns-constants`. 
https://github.com/Zondax/stacks-blockchain-api/blob/440f8bf25ac5a3203ea366e41dcae6ac81cb7162/src/bns-constants.ts#L17

**Comment by zone117x** [comment]
Looks like 2 minor issues need addressed from the review comments above, and this PR needs rebased. Then ready for merge 👍

**Comment by diwakergupta** [comment]
@asimm241 mentioned that this PR also fixes #667 , can we confirm?

https://github.com/blockstack/stacks-blockchain-api/issues/667#issuecomment-882419234

**Review by zone117x — COMMENTED** [review]
Please make sure tests are passing

File: src/datastore/postgres-store.ts (hunk 2 above)
Comment (zone117x) [reply src/datastore/postgres-store.ts:5469]: Looks like a regression after rebase. The above fix was in commit https://github.com/blockstack/stacks-blockchain-api/commit/0ca105278e062f0f7178881e2977006d25b86f3d

**Comment by raynaudoe** [comment]
I'm seeing this error a lot on rosetta's test:
```typescript
    console.error
      Error: Cannot call function on the Dummy datastore. Check if the application is running in offline mode.
          at Proxy.<anonymous> (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/src/datastore/offline-dummy-store.ts:9:13)
          at Object.getRosettaBlockFromDataStore (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/src/api/controllers/db-controller.ts:307:16)
          at /home/runner/work/stacks-blockchain-api/stacks-blockchain-api/src/api/routes/rosetta/network.ts:53:25
          at wrappedMiddleware (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/node_modules/@awaitjs/express/index.js:62:7)
```
What exactly means 'offline mode' ?

**Comment by asimm241** [comment]
> I'm seeing this error a lot on rosetta's test:
> 
> ```ts
>     console.error
>       Error: Cannot call function on the Dummy datastore. Check if the application is running in offline mode.
>           at Proxy.<anonymous> (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/src/datastore/offline-dummy-store.ts:9:13)
>           at Object.getRosettaBlockFromDataStore (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/src/api/controllers/db-controller.ts:307:16)
>           at /home/runner/work/stacks-blockchain-api/stacks-blockchain-api/src/api/routes/rosetta/network.ts:53:25
>           at wrappedMiddleware (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/node_modules/@awaitjs/express/index.js:62:7)
> ```
> 
> What exactly means 'offline mode' ?

This mode is to test rosetta offline. 
see here:  https://www.rosetta-api.org/docs/node_deployment.html#offline-mode-endpoints

that's how you run it in offline mode: https://github.com/blockstack/stacks-blockchain-api/blob/0da61dc5acead2d35c407eee90fac9e94e632f3f/.env#L38

Offline tests: https://github.com/blockstack/stacks-blockchain-api/blob/master/src/tests-rosetta/offline-api-tests.ts

**Comment by raynaudoe** [comment]
Ok, thanks for that info.
Now specifically:

`Cannot call function on the Dummy datastore. Check if the application is running in offline mode.`

refers to the function `getRosettaBlockFromDataStore`, what I don't understand is if I have to add explicitly that function somewhere (mock?) so the dummy-datastore can make us of it ?

**Comment by asimm241** [comment]
Sorry, I don't understand what you are trying to say here but this error is the result of this test case I think. 
https://github.com/blockstack/stacks-blockchain-api/blob/0da61dc5acead2d35c407eee90fac9e94e632f3f/src/tests-rosetta/offline-api-tests.ts#L96

**Comment by zone117x** [comment]
The failing test appears to be `Rosetta API › construction/preprocess - stacking` at
https://github.com/Zondax/stacks-blockchain-api/blob/5785e65fc7ea8b0bba7dd8b8dd3b6f94624ecf3a/src/tests-rosetta/api.ts#L2064
From the log:
```
  ● Rosetta API › construction/preprocess - stacking

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 500

      2062 |       .send(request);
      2063 |
    > 2064 |     expect(result.status).toBe(200);
           |                           ^
      2065 |     expect(result.type).toBe('application/json');
      2066 |
      2067 |     const expectResponse: RosettaConstructionPreprocessResponse = {

      at Object.<anonymous> (tests-rosetta/api.ts:2064:27)
          at runMicrotasks (<anonymous>)

```

**Comment by rllola** [comment]
> The failing test appears to be `Rosetta API › construction/preprocess - stacking` at
> https://github.com/Zondax/stacks-blockchain-api/blob/5785e65fc7ea8b0bba7dd8b8dd3b6f94624ecf3a/src/tests-rosetta/api.ts#L2064
> From the log:
> 
> ```
>   ● Rosetta API › construction/preprocess - stacking
> 
>     expect(received).toBe(expected) // Object.is equality
> 
>     Expected: 200
>     Received: 500
> 
>       2062 |       .send(request);
>       2063 |
>     > 2064 |     expect(result.status).toBe(200);
>            |                           ^
>       2065 |     expect(result.type).toBe('application/json');
>       2066 |
>       2067 |     const expectResponse: RosettaConstructionPreprocessResponse = {
> 
>       at Object.<anonymous> (tests-rosetta/api.ts:2064:27)
>           at runMicrotasks (<anonymous>)
> ```
> 
> Likely related to the error before that:
> 
> ```
>     console.error
>       Error: ABI doesn't contain a function with the name stack_stx
>           at validateContractCall (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/node_modules/@stacks/transactions/src/contract-abi.ts:399:11)
>           at call (/home/runner/work/stacks-blockchain-api/stacks-blockchain-api/node_modules/@stacks/transactions/src/builders.ts:742:5)
> ```

Yeah found the issue. It was me being confused by rosetta operation using `_` and stack function name that use `-`. So when I changed it to fit the standard I also changed the function name.

Another question : Do you also run `test:integration:rosetta` in the CI ? Because I have a bunch of others errors related to `event-stream` folder which is outside of Rosetta scope if I am not mistaken.

**Comment by zone117x** [comment]
> Another question : Do you also run test:integration:rosetta in the CI ? Because I have a bunch of others errors related to event-stream folder which is outside of Rosetta scope if I am not mistaken.

All the tests are ran in CI, and all are now fixed. I was doing final pass through the PR and I'm not really seeing tests for some of the new rosetta stacking operations. Looking at the code coverage, it looks like a couple important sql query results are never returning values (indicating a lack of testing).

See `unlocks` are never returned by the db controller in this function: https://app.codecov.io/gh/blockstack/stacks-blockchain-api/compare/613/diff#D8L512

Or in the sql query here: https://app.codecov.io/gh/blockstack/stacks-blockchain-api/compare/613/diff#D5L5476

Can you add tests for this behavior?

**Comment by rllola** [comment]
> Can you add tests for this behavior?

Sure I can.

I don't have access to codecov.io so I can't see what the link are showing.

**Comment by zone117x** [comment]
> > Can you add tests for this behavior?
> 
> Sure I can.
> 
> I don't have access to codecov.io so I can't see what the link are showing.

I think you'd have to just login to codecov, not sure. But you can always see the code coverage report locally after running tests.

**Comment by rllola** [comment]
@zone117x I am sorry I didn't write this part of the PR. Could you specify what function you want me to add tests for and what behavior I need to check ? Or point me to a similar test that is already written event ?

**Comment by diwakergupta** [comment]
> @zone117x I am sorry I didn't write this part of the PR. Could you specify what function you want me to add tests for and what behavior I need to check ? Or point me to a similar test that is already written event ?

@rllola actually don't worry about it. We'll take over this PR and get it over the finish line, thanks for your help on it thus far! cc/ @asimm241

**Comment by zone117x** [comment]
@asimm241 feel free to pull `Zondax:develop` into a branch in this repo, rebase and squash the commits, then re-open a PR.

**Comment by agraebe** [comment]
closing in favor of https://github.com/blockstack/stacks-blockchain-api/pull/689
